lib/ts/recipe/oauth2/types.ts

```diff
@@ -0,0 +1,298 @@
+              access_token: string;
+              refresh_token?: string;
+              id_token?: string;
+              expires_in: number;
+          }
+        | GeneralErrorResponse
+    >;
+    // The OpenId spec states that we have to support both POST and GET
+    authorizeGET(input: {
+        clientId: string;
+        responseType: ResponseType[];
+        redirectUri: string;
+        scope: string[];
+        state?: string;
+
+        // PKCE
+        codeChallenge?: string;
+        codeChallengeMethod?: CodeChallengeMethod;
+
+        // OpenIdConnect
+        prompt?: OAuthPromptType;
+        nonce?: string;
+
+        session?: SessionContainer;
+
+        options: APIOptions;
+        userContext: any;
+    }): Promise<
+        | {
+              redirectUrl: string; // This will end up as a 302, encoding both error and success
+          }
+        | GeneralErrorResponse
+    >;
+    authorizePOST(input: {
+        clientId: string;
+        responseType: ResponseType[]; // add enum
+        redirectUri: string;
+        scope: string[];
+        state?: string;
+
+        // PKCE
+        codeChallenge?: string;
+        codeChallengeMethod?: CodeChallengeMethod;
+
+        // OpenIdConnect
+        prompt?: OAuthPromptType;
+        nonce?: string;
+
+        session?: SessionContainer;
+
+        options: APIOptions;
+        userContext: any;
+    }): Promise<
+        | {
+              redirectUrl: string; // This will end up as a 302, encoding both error and success
+          }
+        | GeneralErrorResponse
+    >;
+
+    userInfoGET(input: {
+        oauth2AccessToken: string;
+        options: APIOptions;
+        userContext: any;
+    }): Promise<JSONObject | { status: "UNAUTHORISED"; message: string }>;
+};
+
+export type RecipeInterface = {
+    getUserInfo(input: {
+        oauth2AccessToken: string;
+
+        userContext: any;
+    }): Promise<JSONObject>;
+
+    createAuthCode(input: {
+        clientId: string;
+        responseType: ResponseType;
+        redirectUri?: string;
+        scopes: string[];
+
+        // PKCE
+        codeChallenge?: string;
+        codeChallengeMethod?: CodeChallengeMethod;
+
+        // OpenIdConnect
+        prompt?: OAuthPromptType;
+        nonce?: string;
+
+        allQueryParams: string;
+
+        session: SessionContainer;
+        userContext: any;
+    }): Promise<
+        // State is added into the url on the endpoint level
+        | {
+              status: "OK";
+              code?: string;
+              idToken?: string;
+          }
+        | { status: "CLIENT_ERROR"; error: AuthorizationErrorCode; errorDescription: string; errorURI: string } // these are redirected back to the client
+        | { status: "AUTH_ERROR"; error: string } // This gets a redirection to the auth frontend
+    >;
+
+    createTokens(
+        input:
+            | {
+                  clientId: string;
+                  grantType: "authorization_code";
```

Comment:
the core already tries to determine the grant type based on the input. I don't think this is needed in the backend sdk?

Reply:
This will make typing easier when overriding and we already have this information from the API and I planned to expose grant type related `createTokens` variants in the index file to make calling and typing that part easier. (`createTokensWithAuthCode`, etc.)